crates/nebula-importer/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+use quote::__private::Span;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::Write,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+use tokio::{runtime::Builder, sync::Mutex};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(untagged)]
+pub enum Contract {
+    NoDigest(String),
+    WithDigest {
+        digest: Option<String>,
+        reference: String,
+    },
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Config {
+    imports: HashMap<String, Contract>,
+    cache: Option<PathBuf>,
+}
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Metadata {
+    nebula: Config,
+}
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct PackageConfig {
+    pub package: Package,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Package {
+    pub name: String,
+    pub description: Option<String>,
+    pub metadata: Metadata,
+}
+
+impl Contract {
+    fn reference(&self) -> Reference {
+        match self {
+            Contract::NoDigest(reference) => reference.parse().unwrap(),
+            Contract::WithDigest { reference, .. } => reference.parse().unwrap(),
+        }
+    }
+    fn hash(&self) -> String {
+        let mut hasher = Sha256::new();
+        hasher.update(self.reference().to_string().as_bytes());
+        let bytes = hasher.finalize().to_vec();
+        format!("{:x?}", bytes)
+    }
+}
+
+macro_rules! throw_warning {
+    ($($tokens: tt)*) => {
+        println!("cargo:warning={}", format!($($tokens)*))
+    }
+}
+
+pub fn import_all_contracts() {
+    println!("cargo:rerun-if-changed=Cargo.toml");
+    let PackageConfig { package } = Figment::new()
+        .merge(Toml::file("Cargo.toml"))
+        .extract()
+        .expect("Could not read config in `Cargo.toml`.");
+
+    let config = &package.metadata.nebula;
+
+    let contracts_dir = config.cache.clone().unwrap_or({
+        let project_dirs = ProjectDirs::from("co", "eiger", "nebula-importer")
+            .expect("Could not find a base path to cache contracts.");
+        project_dirs.data_local_dir().into()
+    });
+    std::fs::create_dir_all(&contracts_dir)
+        .expect("[importer] Contracts path could not be resolved");
+    sync_contracts(&config, &contracts_dir).expect("[importer] Could not sync contracts.");
+}
+
+pub fn sync_contracts(config: &Config, cache: &PathBuf) -> anyhow::Result<()> {
+    let client = Client::new(oci_distribution::client::ClientConfig {
+        protocol: oci_distribution::client::ClientProtocol::Https,
+        ..Default::default()
+    });
+    let client = Arc::new(Mutex::new(client));
+    let runtime = Builder::new_multi_thread()
+        .worker_threads(1)
+        .enable_all()
+        .build()?;
+
+    for (name, contract) in &config.imports {
+        runtime
+            .block_on(runtime.spawn(find_and_sync_contract(
```

Comment:
AFAIK, this is sequential and will wait until each contract download completes. If thats true, why not avoid `async` dependencies at all here, so the `Arc<Mutex<Client>>` would not be even needed later. Wdyt ?

Reply:
This is not a runtime thing, its build time. 
The oci lib is async so still dont think its posible without writing my own oci implementation.

Reply:
Ok , got it, thanks :)